### ISSUE TITLE
Improve accessibility with screen readers

### ISSUE THREAD

**Issue by avneeshsingh** [paginagmbh/EPUB-Checker#75]
I tried Pagina with JAWS and Windows Narrator on Windows 10, and nothing works.
I am able to hear only the title of the app, nothing else. I also tried to tab to controls but got no response from screen readers.
At this point of time I would say that such a wonderful tool does not work at all with screen readers.

It would be great to see Pagina working with assistive technology. At DAISY Consortium we appreciate the hard work done on Pagina, but lack of accessibility stops us from recommending it.

**Comment by sermo-de-arboribus** [comment]
Thank you @avneeshsingh for this important report. Obviously we've been too comfortable with a Java tutorial statement like this:

> Because support for the Accessibility API is built into the Swing components, your Swing program will probably work just fine with assistive technologies, even if you do nothing special. For example, assistive technologies can automatically get the text information that is set by the following lines of code
(https://docs.oracle.com/javase/tutorial/uiswing/misc/access.html)

Yes, experience tells us that programs or technologies are hardly ever really accessible if the development does not explicitly consider its implications and runs tests for it. Sorry for this negligence. I'm not the person currently in charge at Pagina to continue the EPUB-Checker's development, so I can't promise swift solutions to this problem, but I'll bring this ticket to the attention of the management so they can hopefully prioritize this and apply the requires resources for increasing accessability in the EPUB-Checker app.

**Comment by br-pagina** [comment]
Dear @avneeshsingh ,

I created a new branch on this project for implementing accessibility features:

https://github.com/paginagmbh/EPUB-Checker/tree/dev/75-improve-accessibility

I worked on the following features and have a first version of each of those implemented:

* Make the application accessible to screen readers.
* Make the application usable without mouse, only using the keyboard.
* Implement a high contrast mode.
* Implement options to adjust the font size according to user needs or preferences.

Since your comment explicitly refers to screen readers, let me explain what I did in that area.

* Make it possible to jump to each part of the application. Hitting the tab key now allows you to traverse the different parts previously unreachable without mouse. Standard combinations like "ctrl + shift" allow you to get out of tables, and pressing "F10" opens the top menu.
* Added tool tips to many interactive elements. They are shown at mouse over, and read out by screen readers when the respective element has focus.
* Described components with accessible names and descriptions. These texts only available to screen readers indicate the function and content of UI elements.
* Added shortcuts to open menus: Press "Alt + D" to open the file menu, "Alt + E" for the menu with extracted epubs, "Alt + L" for logs, "Alt + P" for language, "Alt + H" for help. On Mac, "Alt" is replaced by the apple mappings (Ctrl + Option)
* Tested the application using the NVDA screen reader only using the keyboard.

In my tests, I could navigate and use the application using NVDA and keyboard with these changes. However, I would be very interested in your evaluation Would it be possible for you to test this branch of the application and let me know what you think about this? If so, what would you need from me to make this happen?

Best regards,
Bjoern

**Comment by avneeshsingh** [comment]
Thanks @br-pagina.
It would be best if you can provide an test installer for windows. I would also forward it to some other accessibility people who want to see Pagina working for people with accessibility needs.

**Comment by br-pagina** [comment]
Here is a zip file with an executable inside containing the version with the accessibility improvements. You should be able to extract it from the archive and run it on Windows. Please let me know in case there are any problems with running it. [EPUB-Checker.zip](https://github.com/paginagmbh/EPUB-Checker/files/9810744/EPUB-Checker.zip) 
I forgot to mention earlier that at the moment, the localizations for the new features only exist for German and English, since I focused on realizing the functions at this stage. Later, the translations will need to be added.

**Comment by avneeshsingh** [comment]
@br-pagina sorry for the delay. Last month was too busy due to DAISY board meeting. When I did a quick review a month ago, I found it confusing to operate, but I tried it with patience today and was able to see many improvements!

I think most of the labels, menu items  and buttons are spoken now. I tried validating an EPUB, it worked fine. The grid view was a little confusing to use, a screen reader user would need a little time to figure it out.
Then I switched to multiline text report, and it worked well!

I am using JAWS on Windows 10.
We should ask someone to try it out on Mac and VoiceOver also.
I think we can share it more broadly with EPUB community now, and seek wider feedback for accessibility.
Should I forward it to EPUB 3 Working Group and Publishing community group?

**Comment by br-pagina** [comment]
Dear @avneeshsingh ,

thank you for your feedback! 
Good to hear that the changes offer some improvements. I like the idea of making the new version of tool available to a broader range of users so that more feedback can be collected and then implemented. Given that the changes seem to be in the right direction, if you have no objections I would suggest to make the current state a release version that everyone can download and use on their system and with their tools. This should give everyone interested the opportunity to explore and test it and come up with feedback, which I will be happy to implement in the next round.

**Comment by avneeshsingh** [comment]
“Given that the changes seem to be in the right direction, if you have no objections I would suggest to make the current state a release version that everyone can download and use on their system and with their tools.”

 

Please go ahead.
